### PR TITLE
Retry on rate limiting

### LINKED ISSUE #484: XML parsing error when trying to delete lots of objects on s3
When I open a zarr-like array on s3 with `w` mode (i.e., I want to delete the original objects), I occasionally get an xml parsing error:
<details>
<summary>Long traceback </summary>
---------------------------------------------------------------------------
ParseError                                Traceback (most recent call last)
~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _parse_xml_string_to_dom(self, xml_string)
    451             parser.feed(xml_string)
--> 452             root = parser.close()
    453         except XMLParseError as e:

ParseError: no element found: line 2, column 0

During handling of the above exception, another exception occurred:

ResponseParserError                       Traceback (most recent call last)
<ipython-input-13-bad060e4e06e> in <module>
      3     source_arrays = dict(source_group.arrays())
      4     create_success = False
----> 5     target_group, target_arrays = initialize_group(dest_store_path, 
      6                                                source_group.path,
      7                                                arrays=list(source_arrays.values()),

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fibsem_tools/io/io.py in initialize_group(container_path, group_path, arrays, array_paths, chunks, group_attrs, compressor, array_attrs, mode, **kwargs)
    295 ) -> Tuple[zarr.hierarchy.group, Tuple[zarr.core.Array]]:
    296 
--> 297     zgroup = access(
    298         os.path.join(container_path, group_path), mode=mode, attrs=group_attrs, **kwargs
    299     )

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fibsem_tools/io/io.py in access(path, mode, **kwargs)
    157 
    158         if is_container:
--> 159             return accessor(path_outer, path_inner, mode=mode, **kwargs)
    160         else:
    161             return accessor(path_outer, mode=mode, **kwargs)

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fibsem_tools/io/zarr.py in access_n5(dir_path, container_path, **kwargs)
    165 ) -> Any:
    166     dir_path = N5FSStore(dir_path, **kwargs.get("storage_options", {}))
--> 167     return access_zarr(dir_path, container_path, **kwargs)
    168 
    169 

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fibsem_tools/io/zarr.py in access_zarr(dir_path, container_path, **kwargs)
    155         ):
    156             delete_zbranch(tmp)
--> 157     array_or_group = zarr.open(dir_path, path=str(container_path), **kwargs)
    158     if kwargs.get("mode") != "r" and len(attrs) > 1:
    159         array_or_group.attrs.update(attrs)

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/zarr/convenience.py in open(store, mode, **kwargs)
     84             return open_array(store, mode=mode, **kwargs)
     85         else:
---> 86             return open_group(store, mode=mode, **kwargs)
     87 
     88     elif mode == "a":

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/zarr/hierarchy.py in open_group(store, mode, cache_attrs, synchronizer, path, chunk_store, storage_options)
   1169 
   1170     elif mode == 'w':
-> 1171         init_group(store, overwrite=True, path=path, chunk_store=chunk_store)
   1172 
   1173     elif mode == 'a':

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/zarr/storage.py in init_group(store, overwrite, path, chunk_store)
    486 
    487     # initialise metadata
--> 488     _init_group_metadata(store=store, overwrite=overwrite, path=path,
    489                          chunk_store=chunk_store)
    490 

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/zarr/storage.py in _init_group_metadata(store, overwrite, path, chunk_store)
    500     if overwrite:
    501         # attempt to delete any pre-existing items in store
--> 502         rmdir(store, path)
    503         if chunk_store is not None:
    504             rmdir(chunk_store, path)

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/zarr/storage.py in rmdir(store, path)
    120     if hasattr(store, 'rmdir'):
    121         # pass through
--> 122         store.rmdir(path)
    123     else:
    124         # slow version, delete one key at a time

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fibsem_tools/io/storage.py in rmdir(self, path)
    209         store_path = self.dir_path(path)
    210         if self.fs.isdir(store_path):
--> 211             self.fs.rm(store_path, recursive=True)
    212 
    213     def getsize(self, path=None):

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fsspec/asyn.py in wrapper(*args, **kwargs)
     70     def wrapper(*args, **kwargs):
     71         self = obj or args[0]
---> 72         return sync(self.loop, func, *args, **kwargs)
     73 
     74     return wrapper

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fsspec/asyn.py in sync(loop, func, timeout, *args, **kwargs)
     51     event.wait(timeout)
     52     if isinstance(result[0], BaseException):
---> 53         raise result[0]
     54     return result[0]
     55 

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/fsspec/asyn.py in _runner(event, coro, result, timeout)
     18         coro = asyncio.wait_for(coro, timeout=timeout)
     19     try:
---> 20         result[0] = await coro
     21     except Exception as ex:
     22         result[0] = ex

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/s3fs/core.py in _rm(self, path, recursive, **kwargs)
   1508         dirs = [p for p in paths if not self.split_path(p)[1]]
   1509         # TODO: fails if more than one bucket in list
-> 1510         await asyncio.gather(
   1511             *[
   1512                 self._bulk_delete(files[i : i + 1000])

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/s3fs/core.py in _bulk_delete(self, pathlist, **kwargs)
   1496         for path in pathlist:
   1497             self.invalidate_cache(self._parent(path))
-> 1498         await self._call_s3("delete_objects", kwargs, Bucket=bucket, Delete=delete_keys)
   1499 
   1500     async def _rm(self, path, recursive=False, **kwargs):

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/s3fs/core.py in _call_s3(self, method, *akwarglist, **kwargs)
    250             except Exception as e:
    251                 err = e
--> 252         raise translate_boto_error(err)
    253 
    254     call_s3 = sync_wrapper(_call_s3)

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/s3fs/core.py in _call_s3(self, method, *akwarglist, **kwargs)
    231         for i in range(self.retries):
    232             try:
--> 233                 out = await method(**additional_kwargs)
    234                 return out
    235             except S3_RETRYABLE_ERRORS as e:

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/aiobotocore/client.py in _make_api_call(self, operation_name, api_params)
    138             http, parsed_response = event_response
    139         else:
--> 140             http, parsed_response = await self._make_request(
    141                 operation_model, request_dict, request_context)
    142 

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/aiobotocore/client.py in _make_request(self, operation_model, request_dict, request_context)
    158     async def _make_request(self, operation_model, request_dict, request_context):
    159         try:
--> 160             return await self._endpoint.make_request(operation_model, request_dict)
    161         except Exception as e:
    162             await self.meta.events.emit(

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/aiobotocore/endpoint.py in _send_request(self, request_dict, operation_model)
     85         request = await self.create_request(request_dict, operation_model)
     86         context = request_dict['context']
---> 87         success_response, exception = await self._get_response(
     88             request, operation_model, context)
     89         while await self._needs_retry(attempts, operation_model,

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/aiobotocore/endpoint.py in _get_response(self, request, operation_model, context)
    118         # If an exception occurs then the success_response is None.
    119         # If no exception occurs then exception is None.
--> 120         success_response, exception = await self._do_get_response(
    121             request, operation_model)
    122         kwargs_to_emit = {

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/aiobotocore/endpoint.py in _do_get_response(self, request, operation_model)
    178         protocol = operation_model.metadata['protocol']
    179         parser = self._response_parser_factory.create_parser(protocol)
--> 180         parsed_response = parser.parse(
    181             response_dict, operation_model.output_shape)
    182         if http_response.status_code >= 300:

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in parse(self, response, shape)
    243                 parsed = self._do_error_parse(response, shape)
    244         else:
--> 245             parsed = self._do_parse(response, shape)
    246 
    247         # We don't want to decorate event stream responses with metadata

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _do_parse(self, response, shape)
    807         final_parsed['ResponseMetadata'] = self._populate_response_metadata(
    808             response)
--> 809         self._add_modeled_parse(response, shape, final_parsed)
    810         return final_parsed
    811 

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _add_modeled_parse(self, response, shape, final_parsed)
    816         self._parse_non_payload_attrs(response, shape,
    817                                       member_shapes, final_parsed)
--> 818         self._parse_payload(response, shape, member_shapes, final_parsed)
    819 
    820     def _do_modeled_error_parse(self, response, shape):

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _parse_payload(self, response, shape, member_shapes, final_parsed)
    856                     body_shape, original_parsed)
    857         else:
--> 858             original_parsed = self._initial_body_parse(response['body'])
    859             body_parsed = self._parse_shape(shape, original_parsed)
    860             final_parsed.update(body_parsed)

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _initial_body_parse(self, xml_string)
    942         if not xml_string:
    943             return ETree.Element('')
--> 944         return self._parse_xml_string_to_dom(xml_string)
    945 
    946     def _do_error_parse(self, response, shape):

~/miniconda3/envs/cosem-flows/lib/python3.8/site-packages/botocore/parsers.py in _parse_xml_string_to_dom(self, xml_string)
    452             root = parser.close()
    453         except XMLParseError as e:
--> 454             raise ResponseParserError(
    455                 "Unable to parse response (%s), "
    456                 "invalid XML received. Further retries may succeed:\n%s" %

ResponseParserError: Unable to parse response (no element found: line 2, column 0), invalid XML received. Further retries may succeed:
b'<?xml version="1.0" encoding="UTF-8"?>\n'

</details>

The TL;DR is that this operation: [`await self._call_s3("delete_objects", kwargs, Bucket=bucket, Delete=delete_keys)`](https://github.com/dask/s3fs/blob/main/s3fs/core.py#L1576) for a lot of objects (100k+) leads to this exception: `
ResponseParserError: Unable to parse response (no element found: line 2, column 0), invalid XML received. Further retries may succeed:
b'<?xml version="1.0" encoding="UTF-8"?>\n'`

My guess is that this is a side-effect of hitting some API request rate limit, but it seems like that phenomenon should manifest as something a little more explicit than an XML parsing error.

`s3fs==2021.4.0` in case that matters

### PR DESCRIPTION
**Pull request by ian-r-rose**
This treats `ResponseParserErrors` as retryable, as well as `ClientErrors` if they contain `SlowDown`in the message. This can mitigate/fix #484, and is a non-mutually-exclusive alternative to #619. Where that PR tries to address rate limiting at the request time, this tries to address it in the responses.

In my testing with the snippet [here](https://github.com/fsspec/s3fs/issues/484#issuecomment-1113732081), this helps a lot, but is not a complete fix, as I still often exhaust the default retries. So I might be in favor of *both* this and #619 (or something similar to it based on `asyn._run_coros_in_chunks`)

### CODE REVIEW

**Comment by ian-r-rose** [comment]
Okay, I've also added `_run_coros_in_chunks` to accomplish a similar batching to what I did in #619. This appears to work to avoid rate limiting even in the absence of the built-in minimum times that I added over there. So if it would be simpler to consolidate that logic, I'm happy with this approach.

File: s3fs/core.py [hunk 1]
```diff
@@ -12,14 +12,21 @@
 
 from fsspec.spec import AbstractBufferedFile
 from fsspec.utils import infer_storage_options, tokenize, setup_logging as setup_logger
-from fsspec.asyn import AsyncFileSystem, sync, sync_wrapper, FSTimeoutError
+from fsspec.asyn import (
+    AsyncFileSystem,
+    sync,
+    sync_wrapper,
+    FSTimeoutError,
+    _run_coros_in_chunks,
```
Comment (ian-r-rose) [inline s3fs/core.py:20]: I don't love this private import, so I'm given to just vendor this function over here. What do you think @martindurant?

Comment (isidentical) [reply s3fs/core.py:20]: We could even expose it as a public function directly from fsspec.

**Comment by ian-r-rose** [comment]
Yep, that would be best! Though it would require coordinated releases.

**Comment by martindurant** [comment]
S3fs is sister to and closely tied to fsspec - they are always released together. I prefer to keep the function "private" to indicate that ordinary users would not be expected to call it.

**Comment by ian-r-rose** [comment]
Yep, I’m aware they’re both released by you :)

If you’re okay with the cross-repo private import, then I am as well

File: s3fs/core.py [hunk 2]
```diff
@@ -1716,11 +1740,13 @@ async def _rm(self, path, recursive=False, **kwargs):
...
         # TODO: fails if more than one bucket in list
-        await asyncio.gather(
-            *[
+        await _run_coros_in_chunks(
+            [
                 self._bulk_delete(files[i : i + 1000])
                 for i in range(0, len(files), 1000)
-            ]
+            ],
+            batch_size=3,
```
Comment (martindurant) [inline s3fs/core.py:1748]: Where does this batch size come from? I would have thought it could be a much bigger number.

Comment (martindurant) [reply s3fs/core.py:1748]: With this value, how long did it take to delete ~1000 keys or more?

Comment (martindurant) [reply s3fs/core.py:1748]: This is in the context of many dask tasks doing the same in parallel?

Comment (ian-r-rose) [reply s3fs/core.py:1748]: > This is in the context of many dask tasks doing the same in parallel?

No, this is in the context of a bulk delete from a single `S3FileSystem` instance. That occurs in a dask `Client` when `overwrite=True`, but can reproduced by itself as per https://github.com/fsspec/s3fs/issues/484#issuecomment-1113723501

Comment (ian-r-rose) [reply s3fs/core.py:1748]: > these must come back in about 1ms, right?

I think they come back in ~1s, but are batched into bulk delete calls of 1000 keys.

> Now we have the tricky problem that the "best" number will depend on where this is called from.

Yeah, it's tough to come up with a "best" number. That's what I was attempting to do in #619.

Comment (martindurant) [reply s3fs/core.py:1748]: Of course, so sorry - these are thousands of deletes! Please ignore my hesitation in that case. It totally makes sense to send ~3000 requests and wait for them all to end before sending more.

**Comment by martindurant** [comment]
@isidentical , I'm happy with this if you don't have more to say.

File: s3fs/core.py (hunk 2 above)
Comment (ian-r-rose) [reply s3fs/core.py:1748]: I just re-ran my testing scripts with some timing information -- in my case each 1000-key bulk delete took 1-2 seconds, though I don't know how reproducible that is in other contexts.

Comment (martindurant) [reply s3fs/core.py:1748]: +1

**Review by isidentical — APPROVED** [review]
Looks good to me.

**Comment by ian-r-rose** [comment]
Thanks @martindurant!

**Pull request merged**